File: src/main/java/com/braydenoneal/blang/parser/expression/builtin/PlaceBlockBuiltin.java
package com.braydenoneal.blang.parser.expression.builtin;

import com.braydenoneal.blang.parser.Program;
import com.braydenoneal.blang.parser.expression.Expression;
import com.braydenoneal.blang.parser.expression.ExpressionType;
import com.braydenoneal.blang.parser.expression.ExpressionTypes;
import com.braydenoneal.blang.parser.expression.value.*;
import com.mojang.serialization.Codec;
import com.mojang.serialization.MapCodec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.block.Blocks;
import net.minecraft.block.entity.LockableContainerBlockEntity;
import net.minecraft.item.BlockItem;
import net.minecraft.item.ItemStack;
import net.minecraft.item.Items;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;

import java.util.List;

public record PlaceBlockBuiltin(List<Expression> arguments) implements Expression {
    @Override
    public Value<?> evaluate(Program program) {
        Value<?> xValue = arguments.get(0).evaluate(program);
        Value<?> yValue = arguments.get(1).evaluate(program);
        Value<?> zValue = arguments.get(2).evaluate(program);
        Expression itemPredicateExpression = arguments.get(3);

        if (xValue instanceof IntegerValue x &&
                yValue instanceof IntegerValue y &&
                zValue instanceof IntegerValue z &&
                itemPredicateExpression instanceof FunctionValue itemPredicate
        ) {
            BlockPos entityPos = program.context().pos();
            BlockPos pos = new BlockPos(entityPos.getX() + x.value(), entityPos.getY() + y.value(), entityPos.getZ() + z.value());
            World world = program.context().entity().getWorld();

            if (world == null || !world.getBlockState(pos).getBlock().equals(Blocks.AIR)) {
                return null;
            }

            List<LockableContainerBlockEntity> containers = program.context().entity().getConnectedContainers();

            for (LockableContainerBlockEntity container : containers) {
                for (int i = 0; i < container.size(); i++) {
                    ItemStack stack = container.getStack(i);

                    if (stack.isOf(Items.AIR)) {
                        continue;
                    }

                    program.newScope();
                    program.getScope().set(itemPredicate.value().arguments().getFirst(), new ItemValue(stack.getItem()));
                    Value<?> predicateResult = itemPredicate.call(program);
                    program.endScope();

                    if (predicateResult instanceof BooleanValue booleanValue && booleanValue.value()) {
                        for (var entry : BlockItem.BLOCK_ITEMS.entrySet()) {
                            if (stack.isOf(entry.getValue())) {
                                stack.decrement(1);

                                if (stack.isEmpty()) {
                                    container.setStack(i, ItemStack.EMPTY);
                                } else {
                                    container.setStack(i, stack);
                                }

                                world.setBlockState(pos, entry.getKey().getDefaultState());
                            }
                        }

                        return null;
                    }
                }
            }

            return null;
        }

        System.out.println("placeBlock");
        System.out.println(xValue);
        System.out.println(yValue);
        System.out.println(zValue);
        System.out.println(itemPredicateExpression);
        return null;
    }

    public static final MapCodec<PlaceBlockBuiltin> CODEC = RecordCodecBuilder.mapCodec(instance -> instance.group(
            Codec.list(Expression.CODEC).fieldOf("arguments").forGetter(PlaceBlockBuiltin::arguments)
    ).apply(instance, PlaceBlockBuiltin::new));

    @Override
    public ExpressionType<?> getType() {
        return ExpressionTypes.PLACE_BLOCK_BUILTIN;
    }
}

File: src/main/kotlin/com/braydenoneal/blang/parser/expression/builtin/PlaceBlockBuiltin.kt
package com.braydenoneal.blang.parser.expression.builtin

import com.braydenoneal.blang.parser.Program
import com.braydenoneal.blang.parser.RunException
import com.braydenoneal.blang.parser.expression.Arguments
import com.braydenoneal.blang.parser.expression.Expression
import com.braydenoneal.blang.parser.expression.ExpressionType
import com.braydenoneal.blang.parser.expression.ExpressionTypes
import com.braydenoneal.blang.parser.expression.value.BooleanValue
import com.braydenoneal.blang.parser.expression.value.ItemValue
import com.braydenoneal.blang.parser.expression.value.Value
import com.mojang.serialization.MapCodec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.block.Blocks
import net.minecraft.item.BlockItem
import net.minecraft.item.ItemStack
import net.minecraft.item.Items
import net.minecraft.util.math.BlockPos
import java.util.Map


data class PlaceBlockBuiltin(val arguments: Arguments) : Expression {
    override fun evaluate(program: Program): Value<*> {
        val x = arguments.integerValue(program, "x", 0).value
        val y = arguments.integerValue(program, "y", 1).value
        val z = arguments.integerValue(program, "z", 2).value
        val itemPredicate = arguments.functionValue(program, "itemPredicate", 3)

        val entityPos = program.context().pos
        val pos = BlockPos(entityPos.x + x, entityPos.y + y, entityPos.z + z)
        val world = program.context().entity!!.getWorld()

        if (world == null) {
            throw RunException("World is null")
        }

        if (world.getBlockState(pos).block !== Blocks.AIR) {
            return BooleanValue(false)
        }

        val containers = program.context().entity!!.getConnectedContainers()

        for (container in containers) {
            for (slot in 0..<container.size()) {
                val stack = container.getStack(slot)

                if (stack.isOf(Items.AIR)) {
                    continue
                }

                val predicateArguments = Arguments(mutableListOf(ItemValue(stack.item)), Map.of())
                val predicateResult = itemPredicate.call(program, predicateArguments)

                if (predicateResult !is BooleanValue) {
                    throw RunException("itemPredicate is not a predicate")
                }

                if (!predicateResult.value) {
                    continue
                }

                for (entry in BlockItem.BLOCK_ITEMS.entries) {
                    if (!stack.isOf(entry.value)) {
                        continue
                    }

                    stack.decrement(1)

                    if (stack.isEmpty) {
                        container.setStack(slot, ItemStack.EMPTY)
                    } else {
                        container.setStack(slot, stack)
                    }

                    world.setBlockState(pos, entry.key.defaultState)
                    return BooleanValue(true)
                }
            }
        }

        return BooleanValue(false)
    }

    override val type: ExpressionType<*> get() = ExpressionTypes.PLACE_BLOCK_BUILTIN

    companion object {
        val CODEC: MapCodec<PlaceBlockBuiltin> = RecordCodecBuilder.mapCodec {
            it.group(
                Arguments.CODEC.fieldOf("arguments").forGetter(PlaceBlockBuiltin::arguments)
            ).apply(it, ::PlaceBlockBuiltin)
        }
    }
}
